src/main/java/net/minestom/server/entity/EntityProjectile.java

```diff
@@ -152,38 +187,93 @@ private boolean isStuck(Pos pos, Pos posNow) {
             } else {
                 pos = pos.add(direction);
             }
-            Block block = instance.getBlock(pos);
-            if (!block.isAir() && !block.isLiquid()) {
+            if (instance.getBlock(pos).isSolid()) {
                 teleport(pos);
-                return true;
+                return State.StuckInBlock;
+            }
+            if (victimsPredicate == null) {
+                continue;
             }
             Chunk currentChunk = instance.getChunkAt(pos);
             if (currentChunk != chunk) {
                 chunk = currentChunk;
                 entities = instance.getChunkEntities(chunk)
                         .stream()
-                        .filter(entity -> entity instanceof LivingEntity)
+                        .filter(victimsPredicate)
                         .collect(Collectors.toSet());
             }
+
+            final Pos finalPos = pos;
+            Stream<Entity> victims = entities.stream().filter(entity -> entity.getBoundingBox().intersect(finalPos));
+
             /*
-              We won't check collisions with entities for first ticks of arrow's life, because it spawns in the
-              shooter and will immediately damage him.
+              We won't check collisions with self for first ticks of projectile's life, because it spawns in the
+              shooter and will immediately be triggered by him.
              */
             if (getAliveTicks() < 3) {
```

Comment:
In my experience 3 ticks isn't always enough, especially for slower projectiles like splash potions. I think 6 would be a better amount.